File: include/asset/StoredAssetCollection.php
<?php

namespace asset;

use ArrayObject;
use InvalidArgumentException;

/**
 * A collection of `Asset`s.
 * It is used for pagination.
 * @extends ArrayObject<int, StoredAsset>
 */
class StoredAssetCollection extends ArrayObject
{

	/**
	 * @param array<mixed> $assets 
	 * @param null|StoredAssetQuery $nextCollection 
	 * @throws InvalidArgumentException 
	 */
	public function __construct(
		array $assets = array(),
		public ?StoredAssetQuery $nextCollection = NULL
	) {
		foreach ($assets as $asset) {
			if (!$asset instanceof StoredAsset) {
				throw new InvalidArgumentException('All elements must be of type StoredAsset');
			}
		}
		/** @var array<StoredAsset> $assets */
		parent::__construct($assets);
	}

	public function offsetSet(mixed $key, mixed $value): void
	{
		if (!$value instanceof StoredAsset) {
			throw new InvalidArgumentException('All elements must be of type StoredAsset');
		}
		parent::offsetSet($key, $value);
	}

	public function append(mixed $value): void
	{
		if (!$value instanceof StoredAsset) {
			throw new InvalidArgumentException('All elements must be of type StoredAsset');
		}
		parent::append($value);
	}

	/**
	 * Tests whether any of the {@link StoredAsset}s in this collection have the given URL.
	 * Ignoring capitalization.
	 * @param string $url 
	 * @return bool 
	 */
	public function containsUrl(string $url): bool
	{
		foreach ($this as $asset) {
			if (strtolower($asset->url) == strtolower($url)) {
				return true;
			}
		}
		return false;
	}
}

File: app/asset/stored_asset_collection.py
from typing import Iterator, Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from app.asset.stored_asset import StoredAsset
    from app.asset.stored_asset_query import StoredAssetQuery


class StoredAssetCollection:
    def __init__(
        self,
        assets: "list[StoredAsset] | None" = None,
        next_collection: "Optional[StoredAssetQuery]" = None,
    ) -> None:
        self._assets: "list[StoredAsset]" = assets or []
        self.next_collection: "Optional[StoredAssetQuery]" = next_collection

    def append(self, asset: "StoredAsset") -> None:
        self._assets.append(asset)

    def __len__(self) -> int:
        return len(self._assets)

    def __iter__(self) -> "Iterator[StoredAsset]":
        return iter(self._assets)

    def __getitem__(self, index: int) -> "StoredAsset":
        return self._assets[index]

    def contains_url(self, url: str) -> bool:
        url_lower = url.lower()
        return any(a.url.lower() == url_lower for a in self._assets)
